### PR TITLE
RGB Support plus OpenShock Core V1 board definition

### PR DESCRIPTION
**Pull request by nullstalgia**
Includes building binaries for my board!

https://github.com/nullstalgia/OpenShock-Hardware/tree/main/Core

<3

Tweaks to Visual State Manager:

LED during boot is now 3 -> 2 -> 1

3 blinks - WiFi not yet connected
2 blinks - Not yet connected to API
1 blink - All Clear!

When you have GPIO LED + RGB, the GPIO LED is relegated to a simple "All Clear" LED, so we can use RGB for more complex stuff down the line (I plan on adding RF output blinks to the LEDs eventually).

### CODE REVIEW

File: src/RGBPatternManager.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+// Currently this assumes a single WS2812B LED
+// TODO: Support multiple LEDs ?
+// TODO: Support other LED types ?
+
+RGBPatternManager::RGBPatternManager(std::uint8_t rgbPin) : m_rgbPin(rgbPin), m_rgbBrightness(255), m_name {0}, m_pattern(nullptr), m_patternLength(0), m_taskHandle(nullptr), m_taskSemaphore(xSemaphoreCreateBinary()) {
+  snprintf(m_name, sizeof(m_name), "RGBPatternManager-%u", m_rgbPin);
+
+  if ((m_rmtHandle = rmtInit(m_rgbPin, RMT_TX_MODE, RMT_MEM_64)) == NULL) {
+    ESP_LOGE(m_name, "Failed to create rgb rmt object");
```
Comment (hhvrc) [inline src/RGBPatternManager.cpp:19]: Please only use TAG for ESP_LOGX calls, i made this mistake in the past but have corrected it 😅

File: src/RGBPatternManager.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+  }
+
+  float realTick = rmtSetTick(m_rmtHandle, 100);
+  ESP_LOGD(m_name, "real tick set to: %fns", realTick);
```
Comment (hhvrc) [inline src/RGBPatternManager.cpp:23]: Applies to here too

File: src/RGBPatternManager.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+  // Set new values
+  m_patternLength = pattern.size();
+  m_pattern       = new RGBState[m_patternLength];
+  std::copy(pattern.begin(), pattern.end(), m_pattern);
+
+  // Start the task
+  BaseType_t result = xTaskCreate(RunPattern, m_name, 1024, this, 1, &m_taskHandle);
+  if (result != pdPASS) {
+    ESP_LOGE(m_name, "Failed to create task: %d", result);
```
Comment (hhvrc) [inline src/RGBPatternManager.cpp:47]: Applies to here too

File: include/RGBPatternManager.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+#include <nonstd/span.hpp>
+
+#include <freertos/FreeRTOS.h>
+#include <freertos/semphr.h>
+#include <freertos/task.h>
+
+#include <cstdint>
+
+#include <esp32-hal-rmt.h>
```
Comment (hhvrc) [inline include/RGBPatternManager.h:11]: Preferably have this above the <freertos/....h> includes

File: platformio.ini [hunk 5]
```diff
@@ -87,5 +87,16 @@ custom_openshock.chip_variant = N8R8
...
 	-DOPENSHOCK_LED_GPIO=21
 
+; https://github.com/nullstalgia/OpenShock-Hardware/tree/main/Core
+[env:OpenShock-Core-V1]
+board = esp32-s3-devkitc-1 ; builtin
+custom_openshock.chip = ESP32-S3
+custom_openshock.chip_variant = N8R2
+build_flags =
+	-DOPENSHOCK_LED_WS2812B=48
+	-DOPENSHOCK_LED_GPIO=35
```
Comment (hhvrc) [inline platformio.ini:97]: Two leds defined? this is ok as long as it compiles, but idk if this is supported or why we would need it 🤔

**Review by hhvrc — CHANGES_REQUESTED** [review]

File: platformio.ini (hunk 5 above)
Comment (nullstalgia) [reply platformio.ini:97]: I added support for it, the GPIO one just turns into a simple "all is ok" LED, as described in the PR above.

File: include/RGBPatternManager.h (hunk 4 above)
Comment (nullstalgia) [reply include/RGBPatternManager.h:11]: This is autoformatting, not mine.

File: src/RGBPatternManager.cpp [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+void RGBPatternManager::RunPattern(void* arg) {
+  RGBPatternManager* thisPtr = reinterpret_cast<RGBPatternManager*>(arg);
+
+  RGBPatternManager::RGBState* pattern = thisPtr->m_pattern;
+  std::size_t patternLength            = thisPtr->m_patternLength;
+
+  while (true) {
+    for (std::size_t i = 0; i < patternLength; ++i) {
+      thisPtr->SendRGB(pattern[i]);
+      vTaskDelay(pattern[i].duration / portTICK_PERIOD_MS);
```
[the remark this answers is not in the document]
Comment (nullstalgia) [reply src/RGBPatternManager.cpp:133]: Was due to me mis-reading a compiler error during my frustration with the RGB crashes, will undo.

**Comment by nullstalgia** [comment]
So, we seem to have (mostly?) mitigated the RGB crashing on ESP32-S3 boards. But there has not been enough testing, especially for base ESP32 (dual core) and ESP32-C (single core!!) chips.

If there are issues with those, we should consider exploring the RMT Hub option we discussed earlier.

(Also yay I figured out how to get around the issues for my board for now.)

Otherwise, good to go!

**Pull request merged**